lib/Driver/UnixToolChains.cpp

```diff
@@ -356,7 +356,13 @@ std::string toolchains::Android::getTargetForLinker() const {
   return T.str();
 }
 
-bool toolchains::Android::shouldProvideRPathToLinker() const { return false; }
+bool toolchains::Android::shouldProvideRPathToLinker() const {
+#if defined(__ANDROID__)
```

Comment:
This is wrong, the driver should behave the same irrespective of being hosted or cross.

Reply:
The reason it doesn't set an RPATH now is because it assumes cross-compilation, and you don't know what path an app and its Swift shared libraries will be installed to. However, if you're natively compiling on Android, you know the path to libswiftCore.so and so on, so the RPATH should be set for Swift executables and the like.

Reply:
The path of the target environment should be known because we know the layout of android.  So it should be the same everywhere.

Reply:
I'm not getting through, let me go into more detail. There are two possibilities on Android, since it doesn't ship with Swift.

1. You cross-compile the Swift shared libraries, libswiftSwiftOnoneSupport.so and libswiftCore.so, and bundle them with your Android app in its local directory. This is the current default because the Swift compiler _cannot_ know the RPATH, as every app has a different local directory based on its name. That's why it currently doesn't set it.

2. You're natively compiling with Swift in the Termux app. In that case you know exactly where the needed shared libraries are in Termux and would like the Swift compiler to simply add the RPATH for any Swift executables or libraries you build on Android. That's what this patch does. You _may_ not want it to do so if you're natively building Swift shared libraries for an Android app that you then want to install in its own non-Termux path, but you can presumably override the compiler then.

In either case, "the layout of android" is irrelevant, since Android doesn't ship with Swift.

Reply:
The "final intention" is to have an easy-to-use Swift compiler running on Android, building everything from Swift executables to full-blown Android apps __on Android itself__. Termux makes Android one of the most popular linux distros, particularly for users running commands in the shell, as it has between 1 and 5 million downloads on the Play store. Obviously, building a full-blown Android app with its own path in Termux would require overriding and remove this RPATH, but if they're going that far, they know how to do that.

This change is only going to affect those running the Swift compiler on Android, as it's a preprocessor check, and I'm likely the first person to even run the Swift compiler on Android. I'm not sure why you're so worried about how those few people deal with RPATHs. :wink:

Trust me, this RPATH tweak made it easier for me when experimenting with building small Swift executables in Termux.

Reply:
No, you aren't there are others who did actually cross-compile the compiler and run it on android, you are far from the first to do this :-).

Because the behaviour of the compiler being different for the same target based on where/how it is built is very troublesome and makes debugging things far more difficult.

Reply:
>No, you aren't there are others who did actually cross-compile the compiler and run it on android, you are far from the first to do this.

Point me to the evidence.

>Because the behaviour of the compiler being different for the same target based on where/how it is built is very troublesome and makes debugging things far more difficult.

Um, every other Unix platform adds RPATH, only cross-compiling for Android doesn't, for the reasons given above. So for the Swift compiler running on Android to add RPATH would be _more_ consistent.

Reply:
>I will ask you once again to remove this from here.

This is the first time you've asked.

>As we have explain above consistency in the resulting products between different platforms is better.

As I explained above, it all depends on what the "resulting products" is likely to be. If you're cross-compiling to Android, you're much more likely to be packaging Swift shared libraries up in an arbitrary Android app with its own path, so the current disabling of passing rpath makes sense.

However, if you're using the Swift compiler natively on Android, you're more likely to be using it like other Unix platforms, ie building executables or other products to be used within Termux itself, so this patch makes more sense. Since this patch only affects a niche, natively compiling on an Android host, that is currently unused and untested by this project, I'm not sure why you're so worried about it.

>Besides that, adding an rpath is something that can be done from the invocation line of the compiler, while removing this rpath means modifying the result ELF in a later step.

Sure, this is a convenience for the user, as when I first built the Swift compiler and runtime/stdlib libraries on Android, it couldn't find those shared libraries and I had to add them manually.

For the niche of a niche case of building Swift products in Termux that will be added to _another_ Android app with its own path and you want this rpath removed, they can simply extract the appropriate linker command that the Swift compiler's running and remove this rpath before running the modified linker command again. I'd rather enable the common case with this patch, rather than this uncommon case you're worried about.

If all this still doesn't convince you, I can remove this patch. However, it will definitely be patched this way for any mooted Termux package, just external to this repo, and this patch would make life easier for those who want to try compiling the Swift project on Android itself and running its tests there, as I've done.

Reply:
If you want to provide this as a patch in the Termux package build script, it is fine with me.

Reply:
I want to merge it here, for the reasons given above. If you won't merge it here, I'll definitely patch it in the Termux package, but that still won't help people building Swift from scratch on Android, which is the point of this pull.

Reply:
I would really prefer not to have this in the code for the reasons we try to explain above. Baking the rpath to the build-local products is plain wrong.

I have found a new reason. The Android linker (Bionic) doesn’t support RPATH… so I really don’t think this can work as you want. Are you sure this is working?

Reply:
>I would really prefer not to have this in the code for the reasons we try to explain above. Baking the rpath to the build-local products is plain wrong.

I don't understand those reasons, as every other Unix build of the Swift compiler bakes the rpath in non-static executables and shared libraries. This would merely do the same for the Swift compiler running in Termux. As I explained above, the only reason the Android target currently leaves off the rpath is because Swift doesn't ship with Android in /system/lib64, so you'll usually bundle the Swift stdlib with an Android app and place it in an arbitrary app path when cross-compiling. That's _not_ the case when compiling Swift code on Android itself, hence this patch which only affects the Swift compiler when it runs on an Android host.

>I have found a new reason. The Android linker (Bionic) doesn’t support RPATH… so I really don’t think this can work as you want. Are you sure this is working?

Apparently, [DT_RUNPATH is supported by the Bionic linker since Android 7](https://github.com/termux/termux-packages/issues/2071#issuecomment-365771645), which is what the gold linker generates when passed an rpath in the Android 8 device I'm using for testing. [That is one of the reasons the Termux devs are looking at dropping support for Android 5 and 6](https://github.com/termux/termux-packages/blob/master/README.md#android-7-branch-is-in-alpha-testing-) (well, really putting it into maintenance mode with no new packages).

Given the low share of Android 5/6, it's better to support this for Android 7+.

Reply:
Nice that they started supporting RPATH. And if you are fine only supporting Android 7, I suppose you can live with that.

About the topic of RPATH.

CMake builds in one directory, and then installs in another directory. For the full build of Swift, some pieces are build against the Swift in the build directory, and some other later pieces are build against the installation directory (I think products that use swiftpm). The build directory is tied to the machine that builds the code, while the install directory is sometimes for every machine in the platform (but doesn't need to be).

What this option do is adding a `-rpath <swift shared runtime library path>` (this is one path that will probably end in `lib/swift/android/`). When Swift is building, the Swift compiler is being used to compile some parts of the standard library, so those libraries will end up in the build directory with an rpath pointing to the build directory (which is for your machine).

When CMake installs the results into the install directory, it will simply copy the Swift runtime libraries, which will still point to the build directory. If the build directory is still there, the libraries will work. However, if you start using the compiler in the install directory, the rpath of those binaries will point to the install directory. This means that in the case of Swift, `libswiftCore.so` or `libFoundation.so` points to the build directory, but things that build with SwiftPM will point to the install directory. If you remove the build directory, or move the install directory, things start to fail.

Then there's the CMake script, which also adds its own RPATH here and there (using a hardcoded `$ORIGIN:/usr/lib/swift/linux` for Linux, for example, in which `$ORIGIN` is great because is relocatable, but the second part is 🤦‍♂️), and the test script which adds even more.

At some point is not even funny all the patches on top of patches that exist for rpath.

My recommendation: do not switch this. It will only make things worse. If you want to apply a local patch, go for it, I cannot stop you, but please don’t try to submit this here.

Reply:
There are some misconceptions about how rpath is currently used. I just spent some time investigating, here are my results:

>What this option do is adding a -rpath <swift shared runtime library path> (this is one path that will probably end in lib/swift/android/). When Swift is building, the Swift compiler is being used to compile some parts of the standard library, so those libraries will end up in the build directory with an rpath pointing to the build directory (which is for your machine).

No, this doesn't happen, the current CMake config is smarter than this. I haven't dug through all the stdlib CMake config to see where this is set, but the Swift runtime/stdlib doesn't have any DT_RUNPATH set for me on Android. Deleting lib/swift/android/aarch64/libswiftGlibc.so and building it again with ninja -v shows the reason is that CMake calls the local C++ compiler to link that and other Swift libraries, not the Swift compiler itself, so it avoids that potential rpath problem you're worried about.

>If you remove the build directory, or move the install directory, things start to fail.

The former doesn't happen, as the rpath's not set to the build directory.

What you may be thinking of is that it does seem to add the INSTALL_RPATH later in this pull to the Swift libraries' rpath- which I gleaned by checking the shared libraries in the official Swift release for linux, I don't set that install rpath for native builds on Android itself- but since that is separate and set explicitly by the developer to the preferred install lib path, it is a different matter from this rpath invocation by the Swift compiler.

>Then there's the CMake script, which also adds its own RPATH here and there (using a hardcoded $ORIGIN:/usr/lib/swift/linux for Linux, for example, in which $ORIGIN is great because is relocatable, but the second part is man_facepalming), and the test script which adds even more.

I actually don't mind the INSTALL_RPATH being set to /usr/lib/swift/linux on linux, as that could be easily changed by a packager and serves as an example. Perhaps it should be removed for the official Swift release for linux, since that's a standalone build normally run from the user's directory, but not a big deal.

>At some point is not even funny all the patches on top of patches that exist for rpath.

It is true that some extraneous paths have snuck into a few shared libraries' rpaths, here's one from the official Swift 4.2.2 release for Ubuntu 18.04:
```
> readelf -d swift-4.2.2-RELEASE-ubuntu18.04/usr/lib/swift/linux/libXCTest.so | grep RUNPATH
0x000000000000001d (RUNPATH)            Library runpath: [/home/buildnode/jenkins/workspace/oss-swift-4.2-package-linux-ubuntu-18_04/build/buildbot_linux/swift-linux-x86_64/lib/swift/linux:$ORIGIN]
```
While this rpath patch for the Swift compiler may have a role to play in that, it can be done right, as we see most of the main shared libraries shipped with the official Swift release doing.

>My recommendation: do not switch this. It will only make things worse. If you want to apply a local patch, go for it, I cannot stop you, but please don’t try to submit this here.

I understand your concern but I think it's overblown. I'd rather have a Swift compiler that produces working executables as soon as it's built, rather than worry about a few extraneous rpaths here and there. This is how this project configures the Swift compiler for every other non-Windows native toolchain, I see no reason the native Android toolchain should differ. I completely understand why you don't want the rpath for cross-compilation to Android though, which is why this patch keeps rpath disabled there.

I still think this patch should go in for the native Android build of the Swift compiler.

Reply:
In the case of LLVM/Clang, since they also don’t add rpaths to Android, the problem is handled in a local patch to Termux (https://github.com/termux/termux-packages/blob/master/packages/libllvm/tools-clang-lib-Driver-ToolChains-Linux.cpp.patch). Wouldn’t it be better for the Termux community that every compiler is dealt the same? From an outsider point of view, I really think the decisions of a distro/packaging system should not be imposed into the packaged software.

Reply:
Except this patch has nothing to do with the "distro/packaging system"- note that this default rpath patch has nothing to do with Termux, it doesn't explicitly set the Termux sysroot like the clang patch you link- which is why [every other Unix toolchain for Swift sets a default rpath too](https://github.com/apple/swift/blob/master/lib/Driver/UnixToolChains.cpp#L133), as I keep pointing out. Really, it's about "those who want to try compiling the Swift project on Android itself and running its tests there," as I pointed out earlier in this thread, where no Termux packaging patch is going to help you. Now, as I already pointed out, it will _also_ help with a mooted Termux package of the Swift compiler, but I can always add this patch to the Termux package repo for that.

This patch is really about getting the native Android toolchain for Swift in line with every other Unix toolchain for Swift, which will help those building the Swift compiler itself on Android, as I have.